base/hamt.jl

```diff
@@ -66,14 +66,14 @@ mutable struct HAMT{K, V}
     HAMT{K, V}() where {K, V} = new{K,V}(Vector{Union{Leaf{K, V}, HAMT{K, V}}}(undef, 0), zero(BITMAP))
 end
 
-@Base.assume_effects :nothrow function init_hamt(K, V, k, v)
-    # For a single element we can't have a hash-collision
+Base.@assume_effects :nothrow :effect_free function init_hamt(K, V, k, v)
+    # For a single element we can't have a 'hash-collision
     trie = HAMT{K,V}(Vector{Union{Leaf{K, V}, HAMT{K, V}}}(undef, 1), zero(BITMAP))
     trie.data[1] = Leaf{K,V}(k,v)
     return trie
 end
 
-function HAMT{K,V}((k,v)::Pair) where {K, V}
+Base.@assume_effects :effect_free function HAMT{K,V}((k,v)::Pair) where {K, V}
     k = convert(K, k)
     v = convert(V, v)
```

Comment:
Seems fairly sketchy to assert that these convert calls are effect_free

Reply:
Addressed by isolating unsafe cases into a separate function.